src/edge/acrManager.ts

```diff
@@ -0,0 +1,224 @@
+            form: {
+                grant_type: "refresh_token",
+                service: registryUrl,
+                scope,
+                refresh_token: refreshTokenArc,
+            },
+        });
+        return JSON.parse(accessTokenResponse).access_token;
+    }
+
+    private async selectAcrTag(registryUrl: string, repo: string): Promise<vscode.QuickPickItem> {
+        const tag: vscode.QuickPickItem = await vscode.window.showQuickPick(this.loadAcrTagItems(registryUrl, repo), { placeHolder: "Select Tag", ignoreFocusOut: true });
+        return tag;
+    }
+
+    private async loadAcrTagItems(registryUrl: string, repo: string): Promise<vscode.QuickPickItem[]> {
+        try {
+            const accessTokenArc = await this.acquireAccessTokenArc(registryUrl, `repository:${repo}:pull`, this.refreshTokenArc);
+
+            const tagsResponse = await request.get(`https://${registryUrl}/v2/${repo}/tags/list`, {
+                auth: {
+                    bearer: accessTokenArc,
+                },
+            });
+
+            const tagItems: vscode.QuickPickItem[] = [];
+            const tags = JSON.parse(tagsResponse).tags;
+            tags.map((tag) => {
+                tagItems.push({
+                    label: tag,
+                    description: `${registryUrl}/${repo}:${tag}`,
+                });
+            });
+            tagItems.sort((a, b) => a.label.localeCompare(b.label));
+            return tagItems;
+        } catch (error) {
+            error.message = `Error fetching tag list: ${error.message}`;
+            throw error;
+        }
+    }
+
+    private async loginToAcr(registry: Registry, azureSubscription: AzureSubscription) {
+        try {
+            const adminUserEnabled: boolean = registry.adminUserEnabled;
+            const registryUrl: string = registry.loginServer;
+            const loginMessage = adminUserEnabled ? `Looks like the registry "${registryUrl}" has admin user enabled. `
+                + `Would you like to run "docker login" with the admin user credentials?`
+                : `Looks like the registry "${registryUrl}" does not have admin user enabled. `
+                + `Would you like to enable admin user and run "docker login" with the admin user credentials?`;
+            const option: string = await vscode.window.showInformationMessage(loginMessage, "Yes", "No");
+            if (option === "Yes") {
+                vscode.window.withProgress({ title: `Logging in to registry "${registryUrl}"...`, location: vscode.ProgressLocation.Window }, async (progress) => {
+                    const registryName: string = registry.name;
+                    const resourceGroup: string = registry.id.slice(registry.id.toLowerCase().search("resourcegroups/") + "resourcegroups/".length, registry.id.toLowerCase().search("/providers/"));
+                    const client = new ContainerRegistryManagementClient(
```

Comment:
the ContainerRegistryManagementClient should only need one instance after login right?

Reply:
Fixed. Thanks.

Reply:
Sorry, but I reverted this change. `ContainerRegistryManagementClient` is on a subscription basis. Since a user may have multiple subscriptions, there are multiple instance created, and when `ContainerRegistryManagementClient` is [firstly introduced]( https://github.com/Microsoft/vscode-azure-iot-edge/pull/144/files#diff-e8e674adc42104c063931eb583165d99R64), we don't know which one he will pick yet.